packages/next/src/experimental/testmode/playwright/report.ts

```diff
@@ -0,0 +1,104 @@
+import type { TestInfo } from '@playwright/test'
+import type { FetchHandler } from './next-worker-fixture'
+import { step } from './step'
+
+async function parseBody(
+  r: Pick<Request, 'headers' | 'json' | 'text' | 'arrayBuffer' | 'formData'>
+): Promise<Record<string, unknown>> {
+  const contentType = r.headers.get('content-type')
+  let error: string | undefined
+  let text: string | undefined
+  let json: unknown
+  let formData: FormData | undefined
+  let buffer: ArrayBuffer | undefined
+  if (contentType?.includes('text')) {
+    try {
+      text = await r.text()
+    } catch (e) {
+      error = 'failed to parse text'
+    }
+  } else if (contentType?.includes('json')) {
+    try {
+      json = await r.json()
+    } catch (e) {
+      error = 'failed to parse json'
+    }
+  } else if (contentType?.includes('form-data')) {
+    try {
+      formData = await r.formData()
+    } catch (e) {
+      error = 'failed to parse formData'
+    }
+  } else {
+    try {
+      buffer = await r.arrayBuffer()
+    } catch (e) {
+      error = 'failed to parse arrayBuffer'
+    }
+  }
+  return {
+    ...(contentType ? { contentType } : null),
+    ...(error ? { error } : null),
+    ...(text ? { text } : null),
+    ...(json ? { json: JSON.stringify(json) } : null),
+    ...(formData ? { formData: JSON.stringify(Array.from(formData)) } : null),
+    ...(buffer && buffer.byteLength > 0
+      ? { buffer: `base64;${Buffer.from(buffer).toString('base64')}` }
+      : null),
+  }
+}
+
+export async function reportFetch(
+  testInfo: TestInfo,
+  req: Request,
+  handler: FetchHandler
+): Promise<Awaited<ReturnType<FetchHandler>>> {
+  return step(
+    testInfo,
+    {
+      title: `next.onFetch: ${req.method} ${req.url}`,
+      category: 'next.onFetch',
+      apiName: 'next.onFetch',
+      params: {
+        method: req.method,
+        url: req.url,
+        ...(await parseBody(req.clone())),
```

Comment:
Do we want to report the entire body here or just allow parsing that from the response?

Reply:
Report the whole body. The workflow is basically: a developer takes the next unhandled request, copies it's URL and body, and creates a handler for it. Repeat.